# scResolve

**scResolve: Recovering single cell expression profiles from multi-cellular spatial transcriptomics**


<p align="center">
<img src=assets/scresolve.png width="500" height="500">
</p>


scResolve is a deep learning based method that recovers spatially resolved single cell expression profiles from low-resolution spatial transcriptomics and paired histology image.  

It's easy to use. With below 3 command lines, you can obtain an [anndata](https://anndata.readthedocs.io/en/latest/) object storing single-cell gene expression and cell locations in tissue from low-resolution spatial transcriptomics data, like Visium.</br>
``scresolve convert`` - preprocess your data</br>
``scresolve super-resolution `` - run super-resolution to obtain pixel-level expression</br>
``scresolve segment`` - run cell segmentation to obtain expression profiles of single cells and their</br> locations

# Contents
- [Installation](https://github.com/chenhcs/scResolve/tree/main?tab=readme-ov-file#install)
- [Tutorial: Run scResolve on breast cancer dataset from ST platform](https://github.com/chenhcs/scResolve/tree/main?tab=readme-ov-file#quick-version-running-one-slide-of-breast-cancer)
- [Configuration](https://github.com/chenhcs/scResolve/tree/main?tab=readme-ov-file#configuration)
- [Running scResolve on Visium data](https://github.com/chenhcs/scResolve/tree/main?tab=readme-ov-file#running-scresolve-on-your-visium-data)
- [Contact]()



# Installation
scResolve is developed and tested under Python 3.9. Run the following command to install scResolve.
````
conda create -n scresolve python=3.9
conda activate scresolve
pip install git+https://github.com/chenhcs/scResolve@main
````
Once installation is completed, verify the installation by command ``scresolve``.

# Tutorial: Run scResolve on breast cancer dataset from ST platform
Here we provide a tutorial on applying scResolve to the breast cancer dataset generated from the ST platform and reproducing the results from our work.

There are four ST sections in the dataset. Download the data from the link [https://www.spatialresearch.org/resources-published-datasets/doi-10-1126science-aaf2403/](https://www.spatialresearch.org/resources-published-datasets/doi-10-1126science-aaf2403/), or simply use the commands below. We will first use one section as an example, and the same steps can be repeated on the other three sections.

````
curl -Lo section1.jpg https://www.spatialresearch.org/wp-content/uploads/2016/07/HE_layer1_BC.jpg
curl -Lo section1.tsv https://www.spatialresearch.org/wp-content/uploads/2016/07/Layer1_BC_count_matrix-1.tsv
curl -Lo section1-alignment.txt https://www.spatialresearch.org/wp-content/uploads/2016/07/Layer1_BC_transformation.txt
````

The data locations and model parameters are set in the *.toml file [link](https://github.com/chenhcs/scResolve/raw/main/configurations/tutorial.toml), which will be discussed in the next section.

The command below will generate the formatted input for the model.

````
scresolve convert st --counts section1.tsv --image section1.jpg --transformation-matrix section1-alignment.txt --scale 0.3 --save-path section1
````

Next we run super-resolution on the low-resolution spatial transcriptomics and the histology image.

````
scresolve super-resolution tutorial_quick.toml --save-path tutorial
````

This process will generate pixel-level gene expression maps, which will be stored in a tsv file within the ``segment_input`` directory. A grey-scaled jpg image in will be saved in the same path.

Finally, run cell segmentation model.
````
scresolve segment tutorial.toml --count ./segment_input/section1.tsv --image ./segment_input/section1.jpg
````

This step will identify individual cells from the high-resolution expression maps and the histology image. An anndata object storing the expression profiles and locations of cells will be saved.

Results for the other three sections in this dataset can be generated by repeating the same steps. The data can be downloaded using the commands in `get_data.sh`.


# Configuration

scResolve uses below configuration, saved as toml file.

````
[super_resolution_model]
network_depth = 6
network_width = 16
min_counts = 50

[super_resolution_expansion_strategy]
type = "DropAndSplit"
[super_resolution_expansion_strategy.DropAndSplit]
max_metagenes = 50

[super_resolution_optimization]
batch_size = 3
epochs = 100000
learning_rate = 0.0003
patch_size = 768

[super_resolution_analyses]
[super_resolution_analyses.metagenes]
type = "metagenes"
[super_resolution_analyses.metagenes.options]
method = "pca"

[super_resolution_analyses.gene_maps]
type = "gene_maps"
[super_resolution_analyses.gene_maps.options]
gene_regex = ".*"
predict_mean = false
writer = "tensor"

[super_resolution_backtrack]
backtrack = true
min_num_metagenes = 4

[super_resolution_to_segment]
scale_factor = 10

[segment]
prealigned = false
bin_size = 10
ws_otsu_classes = 4
bg_th = 10
epochs = 100
patch_size = 0

[slides]
[slides.section1]
data = "path_to_h5"
image= "path_to_img"
[slides.section1.covariates]
section = 1

[slides.section2]
data = "path_to_h5"
image= "path_to_img"
[slides.section2.covariates]
section = 2
````
There are several important parameters to be considered.  

**Super-resolution model**  
    1. ``backtrack``: this is used for model selection. Set to `true` if you want to backtrack to select a model checkpoint with at least a minimum number of metagenes set by `min_num_metagenes`.  
    2. ``min_num_metagenes``: minimum number of metagenes. Ignored if backtrack is `false`.
    Detailed explanations for other parameters can be found [here](https://github.com/ludvb/xfuse#configuring-and-starting-the-run).

**Cell segmentation model**   
    1. `bg_th`: determines which pixels will be treated as background.  
    2. `ws_otsu_classes`: decides how large the detected nuclei by watershed will be. We recommend using a value of 3 or 4.  

# Running scResolve on Visium data
To run with Visium data, you need to provide paths of below files:  
``bc-matrix``: bc-matrix provided by Visium platform(e.g filtered_feature_bc_matrix.h5).   
``image``: H&E image. We recommend using high resolution image.  
``tissue-position``: Tissue position provided by Visium platform. It has name of tissue_positios_list.csv. Make sure it has a header.  
``scale-factors``: Scale factor provided by Visium platform. A Json format file.  
``mask-file``: We recommend providing a custom mask for super-resolution model. The each pixel in the mask file should be 0 if background, 1 if foreground, 2 if likely background and 4 if likely foreground.  

It is recommended using higher ``scale`` as possible. However, increasing this value requires high GPU resource and will take longer time.  

For the configuration file, you can download default here  

Once you determine above considerations, simply run  

````
scresolve convert visium --bc-matrix PATH --image PATH --tissue-positions PATH --scale-factors PATH --mask-file PATH --scale INT --save-path PATH
````
where each PATH corresponds to the input file.

Then run  
````
scresolve super-resolution CONFIGURATION_FILE --save-path SAVE_PATH
````

Finally, run  
````
scresolve segment CONFIGURATION_FILE --count PATH_TO_SUPER_RESOLUTION_OUTPUT--image PATH_TO_IMAGE_OUTPUT
````

# Contact
Contact us if you have any questions:</br>
Hao Chen: hchen4 at andrew.cmu.edu</br>
Young Je Lee: youngjel at cs.cmu.edu</br>
Jose Lugo-Martinez: jlugomar at andrew.cmu.edu
